### PR TITLE
☣️ [stdlib] Extract and re-use ClosedRange.Index

### PR DESCRIPTION
**Pull request by moiseev**
The type formerly known as `ClosedRange.Index` proves to be quite useful even outside `ClosedRange` (if we remove the requirements imposed on it by it's parent type). One example of such re-use can be found in [apple/swift#14288](https://github.com/apple/swift/pull/14288), where `StrideThrough`'s `Collection` conformance takes advantage of it. Another example is presented here: `LazyPrefixWhileCollection` apparently used a similar type as its `Index`.

I believe this warrants the promotion to the top level, as well as a new name. Please welcome: `PastEndIndex`.

### CODE REVIEW

**Comment by moiseev** [comment]
@swift-ci Please smoke test

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

File: stdlib/public/core/PastEndIndex.swift [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+// FIXME: docs
+public enum PastEndIndex<Bound> {
```
Comment (slavapestov) [inline stdlib/public/core/PastEndIndex.swift]: This type should be @_fixed_layout

Comment (jrose-apple) [reply stdlib/public/core/PastEndIndex.swift]: Aw, and I have to rebase my `@_frozen` PR again. ;-)

**Comment by natecook1000** [comment]
This is a great idea—I've reimplemented this many times.

**Comment by moiseev** [comment]
Compatibility error looks unrelated.

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

**Comment by jrose-apple** [comment]
I'm pretty sure commands are still case-sensitive.

@swift-ci Please test source compatibility

**Comment by moiseev** [comment]
@jrose-apple  https://github.com/apple/swift/pull/15193#issuecomment-372488038 this one worked though...

**Comment by moiseev** [comment]
@swift-ci Please smoke test

**Comment by moiseev** [comment]
Pitched on evolution: https://forums.swift.org/t/extract-the-closedrange-index-type-to-top-level/11711

**Comment by moiseev** [comment]
@swift-ci Please smoke test

**Comment by natecook1000** [comment]
Closing this as out of date — fee free to re-open or continue the discussion in the forums!

**Pull request closed without merging**